src/Compilers/Core/Portable/DiagnosticAnalyzer/DefaultAnalyzerAssemblyLoader.Core.cs

```diff
@@ -121,71 +74,50 @@ public DirectoryLoadContext(string directory, DefaultAnalyzerAssemblyLoader load
             protected override Assembly? Load(AssemblyName assemblyName)
             {
                 var simpleName = assemblyName.Name!;
-                if (_loader.AssemblySimpleNamesToBeLoadedInCompilerContext.Contains(simpleName))
+                try
                 {
-                    // Delegate to the compiler's load context to load the compiler or anything
-                    // referenced by the compiler
-                    return _compilerLoadContext.LoadFromAssemblyName(assemblyName);
+                    if (_compilerLoadContext.LoadFromAssemblyName(assemblyName) is { } compilerAssembly)
+                    {
+                        return compilerAssembly;
+                    }
                 }
-
-                var assemblyPath = Path.Combine(Directory, simpleName + ".dll");
-                var paths = _loader.GetPaths(simpleName);
-                if (paths is null)
+                catch
                 {
-                    // The analyzer didn't explicitly register this dependency. Most likely the
-                    // assembly we're trying to load here is netstandard or a similar framework
-                    // assembly. In this case, we want to load it in compiler's ALC to avoid any 
-                    // potential type mismatch issue. Otherwise, if this is truly an unknown assembly,
-                    // we assume both compiler and default ALC will fail to load it.
-                    return _compilerLoadContext.LoadFromAssemblyName(assemblyName);
+                    // Expected to happen when the assembly cannot be resolved in the compiler / host
+                    // AssemblyLoadContext. 
                 }
 
-                Debug.Assert(paths.Any());
-                // A matching assembly in this directory was specified via /analyzer.
-                if (paths.Contains(assemblyPath))
+                // Prefer registered dependencies in the same directory first.
+                var assemblyPath = Path.Combine(Directory, simpleName + ".dll");
+                if (_loader.IsAnalyzerDependencyPath(assemblyPath))
                 {
-                    return LoadFromAssemblyPath(_loader.GetPathToLoad(assemblyPath));
+                    (_, var loadPath) = _loader.GetAssemblyInfoForPath(assemblyPath);
+                    return LoadFromAssemblyPath(loadPath);
                 }
 
-                AssemblyName? bestCandidateName = null;
-                string? bestCandidatePath = null;
-                // The assembly isn't expected to be found at 'assemblyPath',
-                // but some assembly with the same simple name is known to the loader.
-                foreach (var candidatePath in paths)
+                // Next prefer dependencies that other analyzers registered. Ideally this would not
```

Comment:
I feel like it would be more clear if we just said *Next prefer registered dependencies from other directories*. In some scenarios it is the same analyzer that registered the dependency in msbuild, and in other scenarios it's a different analyzer that registered it.

It might also make sense to note that *this enables analyzer dependencies to be loaded in ProjectReference scenarios, but it also allows packaged analyzers to load which have not properly shipped their dependencies, and are inadvertently relying on dependencies which were shipped by other analyzer packages*.

Eventually I do hope we can eliminate the need for this code path. But it means msbuild needs to stop pointing us at the nuget cache for assemblies that were copied to the output directory.

Reply:
> Eventually I do hope we can eliminate the need for this code path. But it means msbuild needs to stop pointing us at the nuget cache for assemblies that were copied to the output directory.

Which is the bug that outlines this problem? Want to reference it here.